internal/signer/linux/pkcs11/pkcs11.go

```diff
@@ -179,32 +190,30 @@ func (k *Key) Decrypt(encryptedData []byte) ([]byte, error) {
 	return nil, errors.New("decrypt error: Unsupported key type")
 }
 
-func (k *Key) encryptRSA(hash hash.Hash, data []byte) ([]byte, error) {
+func (k *Key) encryptRSA(data []byte) ([]byte, error) {
 	publicKey := k.Public()
 	rsaPubKey := publicKey.(*rsa.PublicKey)
+	hash := cryptoHashToHash(k.hash)
 	return rsa.EncryptOAEP(hash, rand.Reader, rsaPubKey, data, nil)
 }
 
-func (k *Key) encryptRSAWithPKCS11(data []byte) ([]byte, error) {
-	publicKeyFilter := pkcs11.Filter{Class: pkcs11.ClassPublicKey, Label: k.label}
-	pubObjs, err := (k.slot).Objects(publicKeyFilter)
-	if err != nil {
-		return nil, fmt.Errorf("encryptRSAWithPKCS11 error retrieving public key: %v", err)
-	}
-	if len(pubObjs) == 0 {
-		return nil, errors.New("encryptRSAWithPKCS11 error: no public keys found")
-	}
-	pubObj := pubObjs[0]
-	k.privKey = pkcs11.WithPublicKeyHandle(k.privKey, pubObj)
-	return pkcs11.Encrypt(k.privKey, data)
+func (k *Key) decryptRSAWithPKCS11(encryptedData []byte) ([]byte, error) {
+	opts := &rsa.OAEPOptions{Hash: k.hash}
+	return k.decrypter.Decrypt(nil, encryptedData, opts)
 }
 
-func (k *Key) decryptRSAWithPKCS11(encryptedData []byte) ([]byte, error) {
-	return pkcs11.Decrypt(k.privKey, encryptedData)
+func (k *Key) WithHash(hash crypto.Hash) *Key {
+	k.hash = hash
+	return k
 }
 
-func (k *Key) WithHash(hash crypto.Hash) (*Key, error) {
-	res, err := pkcs11.WithHash(k.privKey, hash)
-	k.privKey = res
-	return k, err
+func cryptoHashToHash(hash crypto.Hash) hash.Hash {
+	switch hash {
+	case crypto.SHA256:
+		return sha256.New()
+	case crypto.SHA1:
+		return sha1.New()
+	default:
+		return nil
```

Comment:
This is dangerous and could cause a segfault. Instead this should return a hash and error tuple, to signify there is a failure mode.

Reply:
Ah okay, that makes sense. I'll update it